include/picolibrary/testing/interactive/gpio.h

```diff
@@ -30,6 +30,58 @@
  */
 namespace picolibrary::Testing::Interactive::GPIO {
 
+/**
+ * \brief GPIO input pin state interactive test helper.
+ *
+ * \tparam Output_Stream The type of asynchronous serial output stream to use to output
+ *         information to the user.
+ * \tparam Transmitter The type of asynchronous serial transmitter to use to transmit
+ *         information to the user.
+ * \tparam Input_Pin The type of input pin to toggle.
+ * \tparam Delayer A nullary functor called to introduce a delay each time the pin's state
+ *         is read
+ *
+ * \param[in] transmitter The asynchronous serial transmitter to use to transmit
+ *            information to the user.
+ * \param[in] pin The pin to toggle.
+ * \param[in] delay The nullary functor called to introduce a delay each time the pin's
+ *            state is read.
+ */
+template<template<typename> typename Output_Stream, typename Transmitter, typename Input_Pin, typename Delayer>
+void state( Transmitter transmitter, Input_Pin pin, Delayer delay ) noexcept
+{
+    // #lizard forgives the length
+
+    auto stream = Output_Stream{ std::move( transmitter ) };
+
+    if ( stream.initialize().is_error() ) {
+        return;
+    } // if
+
+    {
+        auto const result = pin.initialize();
+        if ( result.is_error() ) {
+            static_cast<void>( stream.print( "pin initialization error: {}\n", result.error() ) );
+
+            return;
+        } // if
+    }
+
+    for ( ;; ) {
+        delay();
+
+        auto const result = pin.state();
+        if ( result.is_error() ) {
+            static_cast<void>( stream.print( "pin state error: {}\n", result.error() ) );
+
+            return;
+        } // if
+        else {
+            static_cast<void>( result.is_high() ? stream.print( "pin is high\n" ) : stream.print( "pin is low\n" ) );
```

Comment:
This can be accomplished with a single `print()` call using a format string instead of a ternary and two `print()` calls that do not use format strings:
```suggestion
            static_cast<void>( stream.print( "pin is {}\n", result.is_high() ? "high" : "low" ) );
```

In a case where the processing overhead of parsing a format string needed to be avoided for whatever reason, a ternary could be used to select the appropriate string to pass to `put()`:
```suggestion
            static_cast<void>( stream.put( result.is_high() ? "pin is high" : "pin is low" ) );
```

In this case, I do not think that avoiding format string parsing overhead is necessary. Please update to use a call to print with a format string.

Reply:
Going to redo this review to present the findings a bit more coherently.